Allow Structure.interpolate to extrapolate

kyledmiller opened the pull request:
## Summary

Major changes:

- added optional parameter to allow `Structure.interpolate()` to extrapolate beyond the `end_structure`. The parameter is the fractional amount of extrapolation beyond $q=1$ if $q$ is the normalized interpolation mode amplitude. Thus, `extrapolation=0.5` would produce structures spanning $q \in [0,1.5]$. This is helpful for capturing the outer piece of potential energy wells, e.g., when fitting a Landau model to an 2nd order structural phase transition.

## Todos

Create tests

## Checklist

- [x] Google format doc strings added. Check with `ruff`.
- [x] Type annotations included. Check with `mypy`.
- [ ] Tests added for new features/fixes.
- [x] If applicable, new classes/functions/modules have [`duecredit`](https://github.com/duecredit/duecredit) `@due.dcite` decorators to reference relevant papers by DOI ([example](https://github.com/materialsproject/pymatgen/blob/91dbe6ee9ed01d781a9388bf147648e20c6d58e0/pymatgen/core/lattice.py#L1168-L1172))

Tip: Install `pre-commit` hooks to auto-check types and linting before every commit:

```sh
pip install -U pre-commit
pre-commit install
```

janosh commented on pymatgen/core/structure.py:
```diff
@@ -2138,6 +2138,7 @@ def interpolate(
         interpolate_lattices: bool = False,
         pbc: bool = True,
         autosort_tol: float = 0,
+        extrapolation: float = 0,
```
Could you add a test for this new keyword?

kyledmiller commented:
Yep, It's on my to-do list. I'll try to get the tests written by the end of the week.

On Nov 26, 2023, 5:16 PM, at 5:16 PM, Janosh Riebesell ***@***.***> wrote:
>@janosh requested changes on this pull request.
>
>
>
>> @@ -2138,6 +2138,7 @@ def interpolate(
>         interpolate_lattices: bool = False,
>         pbc: bool = True,
>         autosort_tol: float = 0,
>+        extrapolation: float = 0,
>
>Could you add a test for this new keyword?
>
>-- 
>Reply to this email directly or view it on GitHub:
>https://github.com/materialsproject/pymatgen/pull/3467#pullrequestreview-1749446213
>You are receiving this because you authored the thread.
>
>Message ID:
>***@***.***>

kyledmiller commented:
Tests written

janosh reviewed (approved):
Tests look great! Just 1 line length issue left before we can merge.

kyledmiller commented:
Looks good. I fixed that and also added some additions about cell setting & site order consistency to the docstring. If you don't think it belongs or should be in a separate PR, let me know.
